File: core/blas/cgemm.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "cgemm.h"
#include "blas_common.h"
#include "blasplan/include/blasplan/BlasCgemmPlan.h"

using namespace AsdSip;
using namespace Mki;

constexpr int TWO = 2;
constexpr int THREE = 3;
constexpr int64_t PADNUM = 128;

namespace AsdSip {
int64_t GetTrans(const asdBlasOperation_t trans)
{
    int64_t gTrans = -1;
    if (trans == asdBlasOperation_t::ASDBLAS_OP_N) {
        gTrans = 0;
    } else if (trans == asdBlasOperation_t::ASDBLAS_OP_T) {
        gTrans = 1;
    } else if (trans == asdBlasOperation_t::ASDBLAS_OP_C) {
        gTrans = TWO;
    }
    return gTrans;
}

struct CgemmTensorParam {
    const aclTensor *A;
    const aclTensor *B;
    const aclTensor *C;
};

AspbStatus asdBlasCgemmParamCheck(int64_t lda, int64_t ldb, int64_t ldc)
{
    ASDSIP_ECHECK(lda > 0, "blas asdBlasCgemm get lda <= 0.", ErrorType::ACL_ERROR_INVALID_PARAM);
    ASDSIP_ECHECK(ldb > 0, "blas asdBlasCgemm get ldb <= 0.", ErrorType::ACL_ERROR_INVALID_PARAM);
    ASDSIP_ECHECK(ldc > 0, "blas asdBlasCgemm get ldc <= 0.", ErrorType::ACL_ERROR_INVALID_PARAM);
    return ErrorType::ACL_SUCCESS;
}

AspbStatus CgemmDtypeCheck(struct CgemmTensorParam parm)
{
    auto ret = ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE;
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.A, aclDataType::ACL_COMPLEX64, ret);
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.B, aclDataType::ACL_COMPLEX64, ret);
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.C, aclDataType::ACL_COMPLEX64, ret);
    return ErrorType::ACL_SUCCESS;
}

AspbStatus CgemmShapeCheck(struct CgemmTensorParam parm)
{
    auto ret = ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH;
    SIP_OP_CHECK_INVALID_SHAPE(parm.A, ret);
    SIP_OP_CHECK_INVALID_SHAPE(parm.B, ret);
    SIP_OP_CHECK_INVALID_SHAPE(parm.C, ret);
    return ErrorType::ACL_SUCCESS;
}

AspbStatus asdBlasCgemm(asdBlasHandle handle, asdBlasOperation_t transa, asdBlasOperation_t transb, const int64_t m,
    const int64_t n, const int64_t k, const std::complex<float> &alpha, aclTensor *A, const int64_t lda, aclTensor *B,
    const int64_t ldb, const std::complex<float> &beta, aclTensor *C, const int64_t ldc)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    ASDSIP_ECHECK(BlasPlanCache::doesPlanExist(handle),
        "blas Cgemm get cached plan failed.",
        ErrorType::ACL_ERROR_INTERNAL_ERROR);

    struct CgemmTensorParam parm {
        A, B, C
    };
    auto ret = CgemmDtypeCheck(parm);
    ASDSIP_CHECK(ret == ErrorType::ACL_SUCCESS,
        "blas asdBlasCgemm dtype check failed.",
        return ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);

    ret = CgemmShapeCheck(parm);
    ASDSIP_CHECK(ret == ErrorType::ACL_SUCCESS,
        "blas asdBlasCgemm shape check failed.",
        return ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);

    ret = asdBlasCgemmParamCheck(lda, ldb, ldc);
    ASDSIP_CHECK(ret == ErrorType::ACL_SUCCESS,
        "blas asdBlasCgemm param check failed.",
        return ErrorType::ACL_ERROR_INVALID_PARAM);

    try {
        AsdSip::BlasCgemmPlan &plan = dynamic_cast<AsdSip::BlasCgemmPlan &>(BlasPlanCache::getPlan(handle));
        ASDSIP_ECHECK(plan.IsInitialized(), "Cgemm plan init Error!.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

        OpDesc opDesc;
        opDesc.opName = "CgemmOperation";
        AsdSip::OpParam::Cgemm param;

        int64_t gTransA = GetTrans(transa);
        int64_t gTransB = GetTrans(transb);
        int64_t ldaPad = (transa == asdBlasOperation_t::ASDBLAS_OP_N) ? ((m + PADNUM - 1) / PADNUM * PADNUM)
                                                                      : ((k + PADNUM - 1) / PADNUM * PADNUM);
        int64_t ldbPad = (transb == asdBlasOperation_t::ASDBLAS_OP_N) ? ((k + PADNUM - 1) / PADNUM * PADNUM)
                                                                      : ((n + PADNUM - 1) / PADNUM * PADNUM);
        param = {m, n, k, gTransA, gTransB, lda, ldb, ldc, ldaPad, ldbPad, alpha, beta};
        opDesc.specificParam = param;

        if (plan.augAAclTensors.size() != 2 || plan.augBAclTensors.size() != 2 || plan.augCAclTensors.size() != 4) {
            ASDSIP_LOG(ERROR) << "mask tensors error !";
            return ErrorType::ACL_ERROR_INTERNAL_ERROR;
        }
        SVector<aclTensor *> inTensors = {A, B};
        for (auto tensor : plan.augAAclTensors) {
            inTensors.push_back(tensor);
        }
        for (auto tensor : plan.augBAclTensors) {
            inTensors.push_back(tensor);
        }
        for (auto tensor : plan.augCAclTensors) {
            inTensors.push_back(tensor);
        }

        SVector<aclTensor *> outTensors = {C};
        ASDSIP_LOG(DEBUG) << "OpDesc: " << opDesc.opName << "; OpDesc info: " << param.ToString();

        Status cgemmStatus = RunAsdOpsV2(plan.GetStream(), opDesc, inTensors, outTensors, plan.GetWorkspace());
        ASDSIP_ECHECK(cgemmStatus.Ok(), cgemmStatus.Message(), ErrorType::ACL_ERROR_INTERNAL_ERROR);

        ASDSIP_LOG(INFO) << "Execute asdBlasCgemm success.";
        return ErrorType::ACL_SUCCESS;
    } catch (std::bad_cast &e) {
        // Handle the op cgemm exception
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Cgemm Error: " << e.what();
        return ErrorType::ACL_ERROR_INTERNAL_ERROR;
    }
}

AspbStatus asdBlasMakeCgemmPlan(asdBlasHandle handle, asdBlasOperation_t transa, asdBlasOperation_t transb, int64_t m,
    int64_t n, int64_t k, int64_t lda, int64_t ldb, int64_t ldc)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    ASDSIP_ECHECK(handle != nullptr, "blas MakeCgemmPlan failed.", ErrorType::ACL_ERROR_INTERNAL_ERROR);
    ASDSIP_ECHECK(m > 0 && n > 0 && k > 0,
        "blas asdBlasMakeCgemmPlan get m <= 0 || n <= 0 || k <= 0.",
        ErrorType::ACL_ERROR_INVALID_PARAM);

    AsdSip::BlasCgemmPlan *plan = nullptr;
    try {
        plan = new AsdSip::BlasCgemmPlan({transa, transb, m, n, k, lda, ldb, ldc});
        BlasPlanCache::MakePlan(handle, plan);
    } catch (const std::exception &e) {
        if (plan != nullptr) {
            delete plan;
        }
        delete static_cast<int *>(handle);
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Make Cgemm Plan failed: " << e.what();
        throw std::runtime_error("Make Cgemm Plan failed.");
    }
    if (plan->CreateTensor() != ErrorType::ACL_SUCCESS) {
        plan->MarkFailed();
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Fail to create blas asdBlasMakeCgemmPlan mask tensor.";
        return ErrorType::ACL_ERROR_INTERNAL_ERROR;
    }
    plan->MarkInitialized();
    return ErrorType::ACL_SUCCESS;
}
}  // namespace AsdSip

File: sip_pta/test/blas/cgemm.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Copyright (c) 2026 Huawei Technologies Co., Ltd.
This program is free software, you can redistribute it and/or modify it under the terms and conditions of
CANN Open Software License Agreement Version 2.0 (the "License").
Please refer to the License for details. You may not use this file except in compliance with the License.
THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
See LICENSE in the root of the software repository for the full text of the License.
"""

import logging
import random
import sys
from dataclasses import dataclass
from typing import Tuple, Union

import torch
import torch_npu
import torch_sip

# 配置全局日志
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger("torch_sip_cgemm")


@dataclass(frozen=True)
class GemmConfig:
    """参数对象：封装 GEMM 的所有维度和模式配置"""
    m: int
    n: int
    k: int
    trans_a: str = 'N'
    trans_b: str = 'N'
    alpha: Union[complex, float] = 1.0
    beta: Union[complex, float] = 0.0
    name: str = "Case"


class CgemmComprehensiveTester:
    def __init__(self, device: str = "npu:0"):
        self.device = device

    @staticmethod
    def get_random_scalar() -> Union[complex, float]:
        """生成随机标量逻辑"""
        real = random.uniform(-2, 2)
        if random.random() > 0.5:
            imag = random.uniform(-2, 2)
            return complex(real, imag)
        return real

    @staticmethod
    def to_column_major(tensor: torch.Tensor) -> torch.Tensor:
        return tensor.t().contiguous().t()

    def get_complex_mat(self, shape: Tuple[int, ...]) -> torch.Tensor:
        real = torch.randn(shape, dtype=torch.float32)
        imag = torch.randn(shape, dtype=torch.float32)
        return torch.complex(real, imag).to(self.device)

    def run_case(self, cfg: GemmConfig) -> bool:
        # 1. 数据准备 (利用 cfg 属性)
        shape_a = (cfg.k, cfg.m) if cfg.trans_a.upper() != 'N' else (cfg.m, cfg.k)
        shape_b = (cfg.n, cfg.k) if cfg.trans_b.upper() != 'N' else (cfg.k, cfg.n)

        mat_a = self.get_complex_mat(shape_a)
        mat_b = self.get_complex_mat(shape_b)
        mat_c = self.get_complex_mat((cfg.m, cfg.n))
        mat_c_init = mat_c.clone()

        a_in = self.to_column_major(mat_a)
        b_in = self.to_column_major(mat_b)
        c_in = self.to_column_major(mat_c)

        # 2. 核心调用
        try:
            torch_sip.asd_blas_cgemm(
                a_in, b_in, c_in, cfg.alpha, cfg.beta, cfg.trans_a, cfg.trans_b
            )
        except Exception as exc:
            logger.error("[%s] Operator failed: %s", cfg.name, exc)
            return False

        # 3. CPU 参考计算
        def apply_op(m, op):
            op = op.upper()
            return m.t() if op == 'T' else (m.conj().t() if op == 'C' else m)

        a_ref = apply_op(mat_a.cpu(), cfg.trans_a)
        b_ref = apply_op(mat_b.cpu(), cfg.trans_b)
        ref = cfg.alpha * torch.matmul(a_ref, b_ref) + cfg.beta * mat_c_init.cpu()

        # 4. 比对
        is_close = torch.allclose(c_in.cpu(), ref, rtol=1e-3, atol=1e-3)
        if not is_close:
            logger.error("[FAIL] %s | Dim: %dx%dx%d", cfg.name, cfg.m, cfg.n, cfg.k)

        return is_close


def main():
    tester = CgemmComprehensiveTester()
    iteration = 0
    pass_count = 0

    while iteration < 100:
        iteration += 1

        # 使用辅助方法生成随机参数，保持 main 函数整洁
        m, n, k = (random.randint(1, 64) for _ in range(3))

        # 构造配置对象 (解决参数过多问题)
        config = GemmConfig(
            m=m, n=n, k=k,
            alpha=tester.get_random_scalar(),
            beta=tester.get_random_scalar(),
            trans_a=random.choice(['N', 'T', 'C']),
            trans_b=random.choice(['N', 'T', 'C']),
            name=f"Iter_{iteration}"
        )

        if tester.run_case(config):
            pass_count += 1

        if (iteration - 1) % 100 == 0:
            logger.info("Progress: %d/100 cases finished.", iteration)

    return 0 if pass_count == iteration else 1


if __name__ == "__main__":
    sys.exit(main())